### PR TITLE
Ports Motorized Wheelchairs

### PR DESCRIPTION
**Pull request by Bluedazzled**
# Document the changes in your pull request
Original PR: https://github.com/tgstation/tgstation/pull/45737
Ports the best wheelchair technology in Nanotrasen history.
# Wiki Documentation
I don't know about you but I don't think there's any information on wheelchairs in the wiki.
# Changelog
:cl:  TheChosenEvilOne
rscadd: Adds motorized wheelchairs
/:cl:

### CODE REVIEW

File: code/datums/components/crafting/recipes.dm [hunk 1]
```diff
@@ -477,6 +477,19 @@
...
+	name = "Motorized Wheelchair"
+	result = /obj/vehicle/ridden/wheelchair/motorized
+	reqs = list(/obj/item/stack/sheet/metal = 10,
+		/obj/item/stack/rods = 8,
+		/obj/item/stock_parts/manipulator = 2,
+		/obj/item/stock_parts/capacitor = 1)
+	parts = list(/obj/item/stock_parts/manipulator = 2,
+		/obj/item/stock_parts/capacitor = 1)
+	tools = list(TOOL_WELDER, TOOL_SCREWDRIVER, TOOL_WRENCH)
+	time = 200
```
Comment (tattax) [inline code/datums/components/crafting/recipes.dm]:
```suggestion
	time = 20 SECONDS
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+			canmove = FALSE
+			addtimer(VARSET_CALLBACK(src, canmove, TRUE), 20)
+			return FALSE
+		if(power_cell.charge < power_usage / max(power_efficiency, 1))			
+			to_chat(user, "<span class='warning'>The display on [src] blinks 'Out of Power'.</span>")
+			canmove = FALSE
+			addtimer(VARSET_CALLBACK(src, canmove, TRUE), 20)
+			return FALSE
+		if(user.get_num_arms() < arms_required)
+			to_chat(user, "<span class='warning'>You don't have enough arms to operate the motor controller!</span>")
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
			to_chat(user, span_warning("There seems to be no cell installed in [src]."))
			canmove = FALSE
			addtimer(VARSET_CALLBACK(src, canmove, TRUE), 2 SECONDS)
			return FALSE
		if(power_cell.charge < power_usage / max(power_efficiency, 1))			
			to_chat(user, span_warning("The display on [src] blinks 'Out of Power'."))
			canmove = FALSE
			addtimer(VARSET_CALLBACK(src, canmove, TRUE), 2 SECONDS)
			return FALSE
		if(user.get_num_arms() < arms_required)
			to_chat(user, span_warning("You don't have enough arms to operate the motor controller!"))
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		var/atom/throw_target = get_edge_target_turf(H, pick(GLOB.cardinals))
+		unbuckle_mob(H)
+		H.throw_at(throw_target, 2, 3)
+		H.Knockdown(100)
+		H.adjustStaminaLoss(40)
+		if(isliving(M))
+			var/mob/living/D = M
+			throw_target = get_edge_target_turf(D, pick(GLOB.cardinals))
+			D.throw_at(throw_target, 2, 3)
+			D.Knockdown(80)
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
		H.Knockdown(10 SECONDS)
		H.adjustStaminaLoss(40)
		if(isliving(M))
			var/mob/living/D = M
			throw_target = get_edge_target_turf(D, pick(GLOB.cardinals))
			D.throw_at(throw_target, 2, 3)
			D.Knockdown(8 SECONDS)
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+			D.adjustStaminaLoss(35)
+			visible_message("<span class='danger'>[src] crashes into [M], sending [H] and [D] flying!</span>")
+		else
+			visible_message("<span class='danger'>[src] crashes into [M], sending [H] flying!</span>")
+		playsound(src, 'sound/effects/bang.ogg', 50, 1)
+		
+/obj/vehicle/ridden/wheelchair/motorized/emag_act(mob/user)
+	if((obj_flags & EMAGGED) || !panel_open)
+		return
+	to_chat(user, "<span class='warning'>A bomb appears in [src], what the fuck?</span>")
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
	to_chat(user, span_warning("A bomb appears in [src], what the fuck?"))
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+		if(isliving(M))
+			var/mob/living/D = M
+			throw_target = get_edge_target_turf(D, pick(GLOB.cardinals))
+			D.throw_at(throw_target, 2, 3)
+			D.Knockdown(80)
+			D.adjustStaminaLoss(35)
+			visible_message("<span class='danger'>[src] crashes into [M], sending [H] and [D] flying!</span>")
+		else
+			visible_message("<span class='danger'>[src] crashes into [M], sending [H] flying!</span>")
+		playsound(src, 'sound/effects/bang.ogg', 50, 1)
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
			visible_message(span_danger("[src] crashes into [M], sending [H] and [D] flying!"))
		else
			visible_message(span_danger("[src] crashes into [M], sending [H] flying!"))
		playsound(src, 'sound/effects/bang.ogg', 50, TRUE)
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	. += "Speed: [speed]"
+	. += "Energy efficiency: [power_efficiency]"
+	. += "Power: [power_cell.charge] out of [power_cell.maxcharge]"
+
+/obj/vehicle/ridden/wheelchair/motorized/Bump(atom/movable/M)
+	. = ..()
+	// Here is the shitty emag functionality.
+	if(obj_flags & EMAGGED && (istype(M, /turf/closed) || isliving(M)))
+		explosion(src, -1, 1, 3, 2, 0)
+		visible_message("<span class='boldwarning'>[src] explodes!!</span>")
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
		visible_message(span_boldwarning("[src] explodes!!"))
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+						user.visible_message("<span class='notice'>[user] replaces [A] with [B] in [src].</span>", "<span class='notice'>You replace [A] with [B].</span>")
+						break
+			refresh_parts()
+			return
+	return ..()
+
+/obj/vehicle/ridden/wheelchair/motorized/wrench_act(mob/living/user, obj/item/I)
+	to_chat(user, "<span class='notice'>You begin to detach the wheels...</span>")
+	if(I.use_tool(src, user, 40, volume=50))
+		to_chat(user, "<span class='notice'>You detach the wheels and deconstruct the chair.</span>")
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
	to_chat(user, span_notice("You begin to detach the wheels..."))
	if(I.use_tool(src, user, 4 SECONDS, volume = 50))
		to_chat(user, span_notice"You detach the wheels and deconstruct the chair.")
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+			for(var/obj/item/stock_parts/A in contents)
+				for(var/D in required_parts)
+					if(ispath(A.type, D))
+						P = D
+						break
+				if(istype(B, P) && istype(A, P))
+					if(B.get_part_rating() > A.get_part_rating())
+						B.forceMove(src)
+						user.put_in_hands(A)
+						user.visible_message("<span class='notice'>[user] replaces [A] with [B] in [src].</span>", "<span class='notice'>You replace [A] with [B].</span>")
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
		user.visible_message(span_notice("[user] [panel_open ? "opens" : "closes"] the maintenance panel on [src]."), span_notice("You [panel_open ? "open" : "close"] the maintenance panel."))
		return
	if(panel_open)
		if(istype(I, /obj/item/stock_parts/cell))
			if(power_cell)
				to_chat(user, span_warning("There is a power cell already installed."))
			else
				I.forceMove(src)
				power_cell = I
				to_chat(user, span_notice("You install the [I]."))
			refresh_parts()
			return
		if(istype(I, /obj/item/stock_parts))
			var/obj/item/stock_parts/B = I
			var/P
			for(var/obj/item/stock_parts/A in contents)
				for(var/D in required_parts)
					if(ispath(A.type, D))
						P = D
						break
				if(istype(B, P) && istype(A, P))
					if(B.get_part_rating() > A.get_part_rating())
						B.forceMove(src)
						user.put_in_hands(A)
						user.visible_message(span_notice("[user] replaces [A] with [B] in [src]."), span_notice("You replace [A] with [B]."))
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+/obj/vehicle/ridden/wheelchair/motorized/post_unbuckle_mob()
+	. = ..()
+	density = FALSE
+
+/obj/vehicle/ridden/wheelchair/motorized/attack_hand(mob/living/user)
+	if(power_cell && panel_open)
+		power_cell.update_icon()
+		user.put_in_hands(power_cell)
+		power_cell = null
+		to_chat(user, "<span class='notice'>You remove the power cell from [src].</span>")
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
		to_chat(user, span_notice("You remove the [power_cell] from [src]."))
```

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+			return FALSE
+		if(power_cell.charge < power_usage / max(power_efficiency, 1))			
+			to_chat(user, "<span class='warning'>The display on [src] blinks 'Out of Power'.</span>")
+			canmove = FALSE
+			addtimer(VARSET_CALLBACK(src, canmove, TRUE), 20)
+			return FALSE
+		if(user.get_num_arms() < arms_required)
+			to_chat(user, "<span class='warning'>You don't have enough arms to operate the motor controller!</span>")
+			canmove = FALSE
+			addtimer(VARSET_CALLBACK(src, canmove, TRUE), 20)
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
			addtimer(VARSET_CALLBACK(src, canmove, TRUE), 2 SECONDS)
```

File: code/datums/components/crafting/recipes.dm [hunk 11]
```diff
@@ -477,6 +477,19 @@
...
 	time = 10 SECONDS
 	category = CAT_MISC
+	
+/datum/crafting_recipe/motorized_wheelchair
+	name = "Motorized Wheelchair"
+	result = /obj/vehicle/ridden/wheelchair/motorized
+	reqs = list(/obj/item/stack/sheet/metal = 10,
+		/obj/item/stack/rods = 8,
+		/obj/item/stock_parts/manipulator = 2,
+		/obj/item/stock_parts/capacitor = 1)
```
Comment (tattax) [inline code/datums/components/crafting/recipes.dm]:
```suggestion
	reqs = list(/obj/item/stack/sheet/metal = 6,
		/obj/item/stack/rods = 2,
		/obj/item/wheelchair = 1,
		/obj/item/stock_parts/manipulator = 2,
		/obj/item/stock_parts/capacitor = 1)
```

**Comment by tattax** [comment]
> # Wiki Documentation
> No.

yes

File: code/modules/vehicles/motorized_wheelchair.dm [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+						user.visible_message(span_notice("[user] replaces [A] with [B] in [src]."), span_notice("You replace [A] with [B]."))
+						break
+			refresh_parts()
+			return
+	return ..()
+
+/obj/vehicle/ridden/wheelchair/motorized/wrench_act(mob/living/user, obj/item/I)
+	to_chat(user, span_notice("You begin to detach the wheels..."))
+	if(I.use_tool(src, user, 4 SECONDS, volume = 50))
+		to_chat(user, span_notice"You detach the wheels and deconstruct the chair.")
```
Comment (tattax) [inline code/modules/vehicles/motorized_wheelchair.dm]:
```suggestion
		to_chat(user, span_notice("You detach the wheels and deconstruct the chair."))
```

**Comment by JamieD1** [comment]
Have you actually tested this?

**Comment by Bluedazzled** [comment]
Yes I have actually tested this.

**Review by Bibby0110 — APPROVED** [review]

**Pull request merged**